background-processor/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+			check_persisted_data!(nodes[0].node, filepath_0.clone());
+			if done_persisting {
+				loop {
+					if !nodes[0].node.get_persistence_condvar_value() { break }
+				}
+				break
+			}
+		}
+	}
+
+	#[test]
+	fn test_chan_freshness_called() {
+		// Test that ChannelManager's `timer_chan_freshness_every_min` is called every
+		// `CHAN_FRESHNESS_TIMER`.
+		let nodes = create_nodes(1, "test_chan_freshness_called".to_string());
+		let data_dir = nodes[0].persister.get_data_dir();
+		let callback = move |node| FilesystemPersister::persist_manager(data_dir.clone(), node);
+		let processor = BackgroundProcessor::start(callback, nodes[0].node.clone(), nodes[0].logger.clone());
+		loop {
+			let log_entries = nodes[0].logger.lines.lock().unwrap();
+			let desired_log = "Calling manager's timer_chan_freshness_every_min".to_string();
+			if log_entries.get(&("background_processor".to_string(), desired_log)).is_some() {
+				break
+			}
+		}
+		processor.stop();
+		processor.thread_handle.join().unwrap();
+	}
+
+	#[test]
+	fn test_persist_error() {
+		// Test that if we encounter an error during manager persistence, the thread panics.
+		fn persist_manager<ChanSigner, M, T, K, F, L>(_data: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>
+		where ChanSigner: 'static + ChannelKeys + Writeable,
+		      M: 'static + chain::Watch<Keys=ChanSigner>,
+		      T: 'static + BroadcasterInterface,
+		      K: 'static + KeysInterface<ChanKeySigner=ChanSigner>,
+		      F: 'static + FeeEstimator,
+		      L: 'static + Logger,
+		{
+			Err(std::io::Error::new(std::io::ErrorKind::Other, "test"))
+		}
+
+		std::panic::set_hook(Box::new(|panic_info| {
+			if let Some(s) = panic_info.payload().downcast_ref::<String>() {
+				assert_eq!(s, "Errored persisting manager: test");
+			} else {
+				panic!("Expected string panic");
+			}
+		}));
+
+		let nodes = create_nodes(2, "test_persist_error".to_string());
+		let bg_processor = BackgroundProcessor::start(persist_manager, nodes[0].node.clone(), nodes[0].logger.clone());
+		open_channel!(nodes[0], nodes[1], 100000);
+
+		let _ = bg_processor.thread_handle.join().expect_err("Errored persisting manager: test");
+	}
+
+	#[test]
+	fn test_stop_bg_thread() {
+		// Test that we can successfully stop the background thread through BackgroundProcessor::stop().
+		let nodes = create_nodes(1, "test_stop_bg_thread".to_string());
+		let data_dir = nodes[0].persister.get_data_dir();
+		let callback = move |node| FilesystemPersister::persist_manager(data_dir.clone(), node);
+		let bg_processor = BackgroundProcessor::start(callback, nodes[0].node.clone(), nodes[0].logger.clone());
+
+		// Stop the background processor and check that it terminated.
+		bg_processor.stop();
+		loop {
+			let log_entries = nodes[0].logger.lines.lock().unwrap();
+			if log_entries.get(&("background_processor".to_string(), "Terminating background processor.".to_string())).is_some() {
+				break
+			}
+		}
```

Comment:
I'd have to make the `thread_handle` an `Option` to prevent the `cannot move out of X which is behind a shared reference` error when `join`ing in `stop()`. That'd make me wonder whether the `Option` would interfere with users who are also `join()`ing on the thread, which would make me want to add a test. So, I'd rather push this to follow-up unless it seems crucial :P

Reply:
I think the solution is to have `stop` consume `self` (i.e., have it take `self` by value rather than by reference). Then the compiler will enforce the desired behavior.

Reply:
Ah, that helped, thanks. I added this.